dftimewolf/lib/processors/workspace_audit_timesketch.py

```diff
@@ -0,0 +1,175 @@
+          adding to a Timesketch record.
+    """
+    parameters_dict = {}
+    for parameter in parameters:
+      name = parameter.get('name')
+      value = parameter.get('value')
+      if not value:
+        value = parameter.get('multiValue', '')
+        value = ', '.join(value)
+      if name and value:
+        parameters_dict[name] = value
+    return parameters_dict
+
+  def _AddMessageString(self, timesketch_record):
+    """Builds a Timesketch message string from a Timesketch record.
+
+    Args:
+      timesketch_record (dict): a dictionary that will be serialized to JSON
+        and uploaded to Timesketch.
+    """
+    application_name = timesketch_record.get('applicationName')
+    format_strings = self._all_application_format_strings.get(application_name)
+    event_name = timesketch_record.get('_event_name')
+    format_string = format_strings.get(event_name)
+
+    message = ''
+    formatter = string.Formatter()
+    for literal_text, field, _, _ in formatter.parse(format_string):
+      message = message + literal_text
+      if field:
+        if field == 'actor':
+          actor = (timesketch_record.get('actor_email') or
+                   timesketch_record.get('actor_profileId') or
+                   timesketch_record.get('actor_key'))
+          message += '"{0:s}"'.format(actor)
+          continue
+        value = timesketch_record.get(field)
+        if not value:
+          value = timesketch_record.get(field.lower())
+        if not value:
+          value = ''
+        message += '"{0:s}"'.format(value)
+
+    timesketch_record['message'] = message
+
+  def _ProcessLogLine(self, log_record):
+    """Processes a single JSON formatted Google Workspace log line.
+
+    Args:
+      log_record (str): a JSON formatted Workspace log entry.
+
+    Returns:
+      list[str]: one or more Timesketch records.
+    """
+    log_record = json.loads(log_record)
+    actor = self._FlattenActor(log_record.pop('actor', {}))
+    identifiers = log_record.pop('id', {})
+    timestamp = identifiers.pop('time')
+    events = log_record.pop('events', [])
+    timesketch_records = []
+    for event in events:
+      timesketch_record = {
+          'datetime': timestamp,
+          'timestamp_desc': 'Event Recorded',
+          '_event_type': event.get('type'),
+          '_event_name': event.get('name'),
+      }
+      timesketch_record.update(actor)
+      timesketch_record.update(identifiers)
+      timesketch_record.update(log_record)
+
+      parameters = self._FlattenParameters(event.get('parameters', {}))
+      timesketch_record.update(parameters)
+
+      self._AddMessageString(timesketch_record)
+
+      timesketch_records.append(json.dumps(timesketch_record))
+    return timesketch_records
+
+  def _ProcessLogContainer(self, logs_container):
+    """Processes a Workspace logs container.
+
+    Args:
+      logs_container (WorkspaceLogs): logs container.
+    """
+    if not logs_container.path:
+      return
```

Comment:
Should we at least issue a warning here? In which cases will a logs container have an empty `path` attribute?

Reply:
"should never happen"  - i'll add an error message here